### PR TITLE
[NETBEANS-3428] FlatLaf: new close, left, right, drop-down and maximi…

### PR DESCRIPTION
**Pull request by DevCharly**
…ze buttons for multi-tabs, breadcrumb bar, search bar, etc

Old:

image

New:

image

### CODE REVIEW

File: platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/ui/FlatTabControlIcon.java [hunk 1]
```diff
@@ -130,6 +131,21 @@ protected void paintIcon(Component c, Graphics2D g, int width, int height, doubl
         Color bgColor = new Color(0, 0, 0, 0); // Alpha zero means no background.
         Color fgColor = foreground;
         {
+            int buttonState = this.buttonState;
+            if (buttonState == -1) {
+                // get button state from component
+                if (!c.isEnabled()) {
+                    buttonState = TabControlButton.STATE_DISABLED;
```
Comment (eirikbakke) [inline platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/ui/FlatTabControlIcon.java]: Have you tested that this state ends up being used? Are you sure the button does not end up trying to make its own disabled version using an ImageFilter? (In the latter case the disabled version would not look good on HiDPI screens.)

File: platform/openide.awt/src/org/openide/awt/CloseButtonFactory.java [hunk 2]
```diff
@@ -141,6 +141,9 @@ private static boolean isGTKLaF () {
     }
 
     private static Icon getCloseTabImage() {
+        if( null == closeTabImage ) {
+            closeTabImage = UIManager.getIcon("nb.close.tab.icon.enabled.icon"); //NOI18N
```
Comment (eirikbakke) [inline platform/openide.awt/src/org/openide/awt/CloseButtonFactory.java]: Adding new keys to UIManager.getDefaults() constitutes an API change; should probably update apichanges.xml. And perhaps go over the names of the new keys one more time to make sure they follow the convention of the existing keys.

File: platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/FlatLFCustoms.java [hunk 3]
```diff
@@ -45,6 +47,21 @@
             // scale on Java 8 and Linux
             SPLIT_PANE_DIVIDER_SIZE_VERTICAL, UIScale.scale(4),
             SPLIT_PANE_DIVIDER_SIZE_HORIZONTAL, UIScale.scale(4),
+
+            // for org.openide.awt.CloseButtonFactory
```
Comment (eirikbakke) [inline platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/FlatLFCustoms.java]: I remember from https://github.com/apache/netbeans/pull/859/files that the close-button drawing code had to be duplicated between and org.openide.awt.AquaVectorCloseButton and o.n.swing.tabcontrol.plaf.AquaVectorTabControlIcon. Are both of the corresponding locations for FlatLAF updated in this PR? Do the new UIManager keys prevent the need for this duplication?

File: platform/core.multitabs/src/org/netbeans/core/multitabs/ButtonFactory.java [hunk 4]
```diff
@@ -47,7 +48,11 @@
      */
     public static JButton createDropDownButton( Controller controller ) {
         final JButton btn = new JButton( new TabListPopupAction(controller) );
-        btn.setIcon( ImageUtilities.loadImageIcon( "org/netbeans/core/multitabs/resources/down.png", true) ); //NOI18N
+        Icon icon = UIManager.getIcon("nb.multitabs.button.dropdown.icon");
+        if (icon == null) {
+            icon = ImageUtilities.loadImageIcon( "org/netbeans/core/multitabs/resources/down.png", true ); //NOI18N
```
Comment (eirikbakke) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/ButtonFactory.java]: This is just a fallback, right? I assume all of these icons are vector-drawn to work with HiDPI screens?

File: platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/ui/FlatTabControlIcon.java (hunk 1 above)
Comment (DevCharly) [reply platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/ui/FlatTabControlIcon.java]: Yes, this works. The reason is that class `FlatTabControlIcon` is not an instance of `javax.swing.ImageIcon`. So `LookAndFeel.getDisabledIcon()` returns `null` and the normal icon is used.

File: platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/ui/FlatTabControlIcon.java [hunk 5]
```diff
@@ -130,6 +131,21 @@ protected void paintIcon(Component c, Graphics2D g, int width, int height, doubl
         Color bgColor = new Color(0, 0, 0, 0); // Alpha zero means no background.
         Color fgColor = foreground;
         {
+            int buttonState = this.buttonState;
+            if (buttonState == -1) {
```
[the remark this answers is not in the document]
Comment (DevCharly) [reply platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/ui/FlatTabControlIcon.java]: done

File: platform/core.multitabs/src/org/netbeans/core/multitabs/ButtonFactory.java (hunk 4 above)
Comment (DevCharly) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/ButtonFactory.java]: No, not just a fallback.
Vector icons are only used for FlatLaf.

Other LaFs still use the bitmap images.

Comment (DevCharly) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/ButtonFactory.java]: But should be not very difficult to extend `Windows8VectorTabControlIcon` and `AquaVectorTabControlIcon` in the same way I've extended `FlatTabControlIcon`...

File: platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/FlatLFCustoms.java (hunk 3 above)
Comment (DevCharly) [reply platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/FlatLFCustoms.java]: Yes, I thinks so. Also FlatLaf is a single module, but Aqua/Windows LAF support is scattered to many modules.
Anyway, there is only one close button in FlatLaf in `FlatTabControlIcon`.

File: platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/ui/FlatTabControlIcon.java (hunk 1 above)
Comment (eirikbakke) [reply platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/ui/FlatTabControlIcon.java]: Great!

File: platform/openide.awt/src/org/openide/awt/CloseButtonFactory.java (hunk 2 above)
Comment (lkishalmi) [reply platform/openide.awt/src/org/openide/awt/CloseButtonFactory.java]: While introducing new keys is an API change, the keys are not documented in ```apichange.xml```. That could be an additional task, or we could introduce a LAF guide on wiki. I would not hold this one off due to that.

Comment (eirikbakke) [reply platform/openide.awt/src/org/openide/awt/CloseButtonFactory.java]: Sure--it's not worth any extra effort now.

**Pull request merged**